### PR TITLE
Concurrent Scavenger Ownable Synchronizer Fix

### PR DESCRIPTION
**Pull request by RSalman**
Fix corner case where Ownable Sync Objet is duplicated in the Ownable Sync list. This occurs
when Ownable object construction finishes much after allocation causing
it to survive multiple scavenges and being tenured. The object will be
added by jit following the STW phase, the concurrent phase will
attempt to add it again. The current `!isObjectInNewSpace(link)` condition would
stop this in normal circumstances. However, since this object will be
linked to tenure, `!isObjectInNewSpace(link)` will not be a sufficient
check as it only checks against survivor space which would cause the tenured Ownable to pass
the condition and be added again. 

Signed-off-by: Salman Rana <salman.rana@ibm.com>

### CODE REVIEW

**Comment by RSalman** [comment]
@dmitripivkine @amicic please review

**Comment by RSalman** [comment]
_Trace of Ownable Object being duplicated:_

```
*** AllocateObject [00000000FFF224D0] [inlineAllocateObject]: [java/util/concurrent/locks/ReentrantReadWriteLock$FairSync] ***
COPY [00000000FFF224D0 -> 00000000FFF91100] [java/util/concurrent/locks/ReentrantReadWriteLock$FairSync] // Flipped
MM_ScavengerDelegate::private_addOwnableSynchronizerObjectInList LINK == NULL [object: 00000000FFF91100] // Not added, link is null
COPY [00000000FFF91100 -> 00000000FFF11198] [java/util/concurrent/locks/ReentrantReadWriteLock$FairSync] // Flipped
MM_ScavengerDelegate::private_addOwnableSynchronizerObjectInList LINK == NULL [object: 00000000FFF11198] // Not added, link is null
COPY [00000000FFF11198 -> 000000005F765BB8] [java/util/concurrent/locks/ReentrantReadWriteLock$FairSync] // Tenured 
Buffer Add: [0] ObjCreated -> 000000005F765BB8 [REGION: 000000005B600000-00000000647F0000] [count: 1] //Allocated Ownable is finally added
MM_ScavengerDelegate::private_addOwnableSynchronizerObjectInList [object: 000000005F765BB8] [Link: 000000005F765BB8] // Tenured address seen by scavenger
Buffer Add: ScavengerAddOwnable [isConcurrentCycleInProgress & isCurrentPhaseConcurrent] -> 000000005F765BB8 [REGION: 000000005B600000-00000000647F0000] [count: 1]
```

_Full Log_

```
<concurrent-start>
	{SCAV: masterThreadConcurrentCollect:: About to dispatch}
<concurrent-end>

<exclusive-start>
<af-start>
	<STW 2 phase Start>
		masterThreadGarbageCollect START -> Check Ownable List [Concurrent State: 4] 
	<Cycle END>
	<CYCLE START - Scavenger STW 1 Start>
		masterThreadGarbageCollect START -> Check Ownable List [Concurrent State: 0] 
	<Scavenger STW 1 END>
<af-end>
<exclusive-end>

<concurrent-start>
	{SCAV: masterThreadConcurrentCollect:: About to dispatch}
    *** AllocateObject [00000000FFF224D0] [inlineAllocateObject]: [java/util/concurrent/locks/ReentrantReadWriteLock$FairSync] ***
<concurrent-end>



<exclusive-start>
<af-start>
	<STW 2 Start>
		masterThreadGarbageCollect START -> Check Ownable List [Concurrent State: 4]
	<Cycle END - STW 2 end>
<af-end>
<exclusive-end>


<exclusive-start>
<af-start>
	<CYCLE START - Scavenger STW 1 Start>
		masterThreadGarbageCollect START -> Check Ownable List [Concurrent State: 0]
		*** [0] COPY [00000000FFF224D0 -> 00000000FFF91100] [java/util/concurrent/locks/ReentrantReadWriteLock$FairSync] *** 
	<Scavenger STW 1 END>
<af-end>
<exclusive-end>

<exclusive-start>
<af-start>
	<Concurrent phase in STW / STW 2 phase Start>
		masterThreadGarbageCollect START -> Check Ownable List [Concurrent State: 3] 
		scavengeScan - about to dispatch
		[0] MM_ScavengerDelegate::private_addOwnableSynchronizerObjectInList LINK == NULL [object: 00000000FFF91100] 
	<Cycle END>
	<CYCLE START - Scavenger STW 1 Start>
		masterThreadGarbageCollect START -> Check Ownable List [Concurrent State: 0]
		*** [0] COPY [00000000FFF91100 -> 00000000FFF11198] [java/util/concurrent/locks/ReentrantReadWriteLock$FairSync] *** 
	<Scavenger STW 1 END>
<af-end>
<exclusive-end>

<concurrent-start>
	{SCAV: masterThreadConcurrentCollect:: About to dispatch}
	[0] MM_ScavengerDelegate::private_addOwnableSynchronizerObjectInList LINK == NULL [object: 00000000FFF11198] 
<concurrent-end>

<exclusive-start>
<af-start>
	<STW 2 Start>
		masterThreadGarbageCollect START -> Check Ownable List [Concurrent State: 4] 
	<Cycle END - STW 2 end>
<af-end>
<exclusive-end>

<exclusive-start>
<af-start>
	<CYCLE START - Scavenger STW 1 Start>
		masterThreadGarbageCollect START -> Check Ownable List [Concurrent State: 0]
		*** [0] COPY [00000000FFF11198 -> 000000005F765BB8] [java/util/concurrent/locks/ReentrantReadWriteLock$FairSync] *** 
	<Scavenger STW 1 END>
<af-end>
<exclusive-end>

Buffer Add: [0] ObjCreated -> 000000005F765BB8 [REGION: 000000005B600000-00000000647F0000] [count: 1]

<concurrent-start>
	{SCAV: masterThreadConcurrentCollect:: About to dispatch}
	[0] MM_ScavengerDelegate::private_addOwnableSynchronizerObjectInList [object: 000000005F765BB8] [Link: 000000005F765BB8] 
	[0] Buffer Add: ScavengerAddOwnable [isConcurrentCycleInProgress & isCurrentPhaseConcurrent] -> 000000005F765BB8 [REGION: 000000005B600000-00000000647F0000] [count: 1]
```

**Comment by amicic** [comment]
jenkins test sanity win jdk8

**Comment by amicic** [comment]
jenkins test extended xLinux jdk8

File: runtime/gc_glue_java/ScavengerDelegate.cpp [hunk 1]
```diff
@@ -604,12 +604,12 @@ MM_ScavengerDelegate::private_addOwnableSynchronizerObjectInList(MM_EnvironmentS
...
 	 * it could be the constructing object which would be added in the list after the construction finish later. ignore the object to avoid duplicated reference in the list.
-	 * For concurrent scavenger, an object that doesn't finish constructing before the start of the STW phase will be added to the list after, during the concurrent phase.
-	 * In this case, the object may already be added to the list. */
-	if (NULL != link && (!_extensions->isConcurrentScavengerEnabled() || (_extensions->isConcurrentScavengerEnabled() && !_extensions->scavenger->isObjectInNewSpace(link)))) {
+	 * For concurrent scavenger, an object that doesn't finish constructing before the start of the STW phase will be added to the list after. As a result, the object may already
+	 * be added to the list (non NULL link) when we scan it during the concurrent phase. We must not add it again here, so check link before proceeding */
+	if (NULL != link && (!_extensions->isConcurrentScavengerEnabled() || (_extensions->isConcurrentScavengerEnabled() && _extensions->scavenger->isObjectInEvacuateMemory(link)))) {
```
Comment (dmitripivkine) [inline runtime/gc_glue_java/ScavengerDelegate.cpp:609]: please add extra pair of brackets `(NULL != link)`

**Comment by amicic** [comment]
jenkins test extended xLinux jdk8

**Pull request merged**